internal/graph/storagewrapper.go

```diff
@@ -0,0 +1,340 @@
+	options storage.ReadOptions,
+) (storage.TupleIterator, error) {
+	ctx, span := tracer.Start(
+		ctx,
+		"cache.Read",
+		trace.WithAttributes(attribute.Bool("cached", false)),
+	)
+	defer span.End()
+
+	iter := func(ctx context.Context) (storage.TupleIterator, error) {
+		return c.OpenFGADatastore.Read(ctx, store, tupleKey, options)
+	}
+
+	if options.Consistency.Preference == openfgav1.ConsistencyPreference_HIGHER_CONSISTENCY {
+		return iter(ctx)
+	}
+
+	var b strings.Builder
+	b.WriteString(
+		fmt.Sprintf("%sr%s/%s", QueryCachePrefix, store, tuple.TupleKeyToString(tupleKey)),
+	)
+	return c.newCachedIterator(ctx, iter, b.String())
+}
+
+// newCachedIterator either returns a cached static iterator for a cache hit, or
+// returns a new iterator that attempts to cache the results.
+func (c *CachedDatastore) newCachedIterator(
+	ctx context.Context,
+	dsIterFunc iterFunc,
+	key string,
+) (storage.TupleIterator, error) {
+	span := trace.SpanFromContext(ctx)
+	tuplesCacheTotalCounter.Inc()
+
+	cacheKey := cacheKeyFor(key)
+
+	if cacheKey == "" {
+		return dsIterFunc(ctx)
+	}
+
+	cachedResp := c.cache.Get(cacheKey)
+	isCached := cachedResp != nil && !cachedResp.Expired && cachedResp.Value != nil
+
+	if isCached {
+		tuplesCacheHitCounter.Inc()
+		span.SetAttributes(attribute.Bool("cached", true))
+		return storage.NewStaticTupleIterator(cachedResp.Value.([]*openfgav1.Tuple)), nil
+	}
+
+	iter, err := dsIterFunc(ctx)
+	if err != nil {
+		return nil, err
+	}
+
+	return &cachedIterator{
+		iter:          iter,
+		tuples:        make([]*openfgav1.Tuple, 0, c.maxResultSize),
+		cacheKey:      cacheKey,
+		cache:         c.cache,
+		maxResultSize: c.maxResultSize,
+		ttl:           c.ttl,
+	}, nil
+}
+
+// Close closes the datastore and cleans up any residual resources.
+func (c *CachedDatastore) Close() {
+	c.OpenFGADatastore.Close()
+}
+
+type cachedIterator struct {
+	iter     storage.TupleIterator
+	tuples   []*openfgav1.Tuple
+	cacheKey string
+	cache    storage.InMemoryCache[any]
+	ttl      time.Duration
+
+	// maxResultSize is the maximum number of tuples to cache. If the number
+	// of tuples found exceeds this value, it will not be cached.
+	maxResultSize int
+
+	// sf is used to prevent draining the same iterator
+	// across multiple requests.
+	sf singleflight.Group
+
+	// closeOnce is used to synchronize .Close() and ensure
+	// it's only done once.
+	closeOnce sync.Once
+
+	// wg is used purely for testing and is an internal detail.
+	wg sync.WaitGroup
+}
+
+// Next see [Iterator.Next].
+func (c *cachedIterator) Next(ctx context.Context) (*openfgav1.Tuple, error) {
+	t, err := c.iter.Next(ctx)
+	if err != nil {
+		if !errors.Is(err, storage.ErrIteratorDone) &&
+			!errors.Is(err, context.Canceled) &&
+			!errors.Is(err, context.DeadlineExceeded) {
+			c.tuples = nil // don't store results that are incomplete
+		}
+		return nil, err
+	}
+
+	c.addToBuffer(t)
+
+	return t, nil
+}
+
+// Stop see [Iterator.Stop].
```

Comment:
could you improve this to say that this function runs the iterator til the very end in the background?

Reply:
updated comments for .Next() and .Stop()